pkg/catalog/catalog.go

```diff
@@ -161,21 +174,49 @@ func newSharedCatalog() (*Catalog, error) {
 	return c, nil
 }
 
-// getCatalogCache retrieves the catalog from the local directory.
-func getCatalogCache() (catalog *Catalog, err error) {
-	b, err := os.ReadFile(getCatalogCachePath())
+// getCatalogCache retrieves the catalog from the local directory along with locking the catalog file
+// If `setWriteLock` is false, it will read the catalog file with ReadLock and release the lock at the same time
+// If `setWriteLock` is true, it will apply WriteLock to the catalog file, read the catalog file
+// and keep the WriteLock to the file along with returning `unlock` function. It is caller
+// functions responsibility to unlock the WriteLock after the catalog update
+func getCatalogCache(setWriteLock bool) (catalog *Catalog, unlock func(), err error) {
+	unlock = func() {}
+	var b []byte
+
+	if setWriteLock {
+		var lockedFile *lockedfile.File
+		if utils.PathExists(getCatalogCachePath()) {
+			lockedFile, err = lockedfile.Edit(getCatalogCachePath())
+		} else {
+			// Create directory path if missing before locking the file
+			_ = os.MkdirAll(filepath.Dir(getCatalogCachePath()), 0777)
+			lockedFile, err = lockedfile.Create(getCatalogCachePath())
+		}
+		if err != nil {
+			return nil, nil, err
+		}
+		unlock = func() {
+			if lockedFile != nil {
+				lockedFile.Close()
+			}
+		}
+		b, err = io.ReadAll(lockedFile)
```

Comment:
If this fail with an error, we should not fall out and create a `newSharedCatalog()`, since it will not be locked.
An error here is a real error and we should return.

Reply:
I did not get what you mean by `since it will not be locked`? The lock is acquired when we do `lockedFile, err = lockedfile.Edit(getCatalogCachePath())` right?

So, if the `ReadAll` fails the lock is already acquired and we will create a new sharedCatalog.

This is no different from the below line where we do `lockedfile.Read(getCatalogCachePath())` which internally does
```
Open()
io.ReadAll(lockedFile)
Close()
```

Reply:
You are right. My comment about “not being locked” doesn’t make sense. 

The risk however is that if the readAll fails, we’ll overwrite the existing catalog with a new empty one.  But I don’t know what could make the read fail…

But you are right that it’s the same as the line `lockedfile.Read(getCatalogCachePath())`.  The code assumes that if there is an error with the read it’s because the file doesn’t exist and therefore we create a new empty catalog. But what if the read fails for some other reason?  Maybe this could happen with a lock timeout?

Reply:
> But you are right that it’s the same as the line lockedfile.Read(getCatalogCachePath()). The code assumes that if there is an error with the read it’s because the file doesn’t exist and therefore we create a new empty catalog. But what if the read fails for some other reason? Maybe this could happen with a lock timeout?

Yeah makes sense. I have updated the code to check the type of error and only create the `newSharedCatalog` if `os.IsNotExist(err)` is `true`.